### PR TITLE
Canny custom gradient

### PR DESCRIPTION
**Pull request by alalek**
Related #6804

### CODE REVIEW

File: modules/imgproc/src/canny.cpp [hunk 1]
```diff
@@ -53,61 +57,82 @@
 
 namespace cv
 {
+
+static void CannyImpl(Mat& dx_, Mat& dy_, Mat& _dst, double low_thresh, double high_thresh, bool L2gradient);
+
+
 #ifdef HAVE_IPP
-static bool ippCanny(const Mat& _src, Mat& _dst, float low,  float high)
+template <int useCustomDeriv>
```
Comment (ilya-lavrenov) [inline modules/imgproc/src/canny.cpp]: why not `bool`?

File: modules/imgproc/src/canny.cpp [hunk 2]
```diff
@@ -685,14 +691,60 @@ while (borderPeaks.try_pop(m))
...
+    }
+
+    CannyImpl(dx, dy, dst, low_thresh, high_thresh, L2gradient);
+}
+
+static void CannyImpl(Mat& dx, Mat& dy, Mat& dst,
+    double low_thresh, double high_thresh, bool L2gradient)
+{
+    const int cn = dx.channels();
+    const int cols = (int)dx.cols, rows = (int)dx.rows;
```
Comment (ilya-lavrenov) [inline modules/imgproc/src/canny.cpp]: do we need casting here?

File: modules/imgproc/src/canny.cpp [hunk 3]
```diff
@@ -62,60 +62,77 @@ static void CannyImpl(Mat& dx_, Mat& dy_, Mat& _dst, double low_thresh, double h
 
 
 #ifdef HAVE_IPP
-static bool ippCanny(const Mat& _src, Mat& _dst, float low,  float high)
+template <bool useCustomDeriv>
+static bool ippCanny(const Mat& _src, const Mat& dx_, const Mat& dy_, Mat& _dst, float low, float high)
```
Comment (ilya-lavrenov) [inline modules/imgproc/src/canny.cpp]: we can add an internal assert that either `_src` or both `dx_` and `dy_` must be empty.

File: modules/imgproc/src/canny.cpp [hunk 4]
```diff
@@ -685,14 +717,66 @@ while (borderPeaks.try_pop(m))
...
 
+    CannyImpl(dx, dy, dst, low_thresh, high_thresh, L2gradient);
+#endif
+}
+
+void Canny( InputArray _dx, InputArray _dy, OutputArray _dst,
+                double low_thresh, double high_thresh,
+                bool L2gradient )
+{
+    CV_Assert(_dx.dims() == 2 && _dy.dims() == 2);
```
Comment (ilya-lavrenov) [inline modules/imgproc/src/canny.cpp]: we can check only `_dx.dims() == 2`. For `_dy` the condition will be checked at the line 747

**Comment by vpisarev** [comment]
:+1:

**Pull request merged**